src/rust/container/lib.rs

```diff
@@ -0,0 +1,384 @@
+                    receiver: input_receiver,
+                };
+                let output_stream = AsyncWriter {
+                    sender: output_sender,
+                };
+                let network = capnp_rpc::twoparty::VatNetwork::new(
+                    input_stream,
+                    output_stream,
+                    rpc_twoparty_capnp::Side::Server,
+                    Default::default(),
+                );
+                let rpc_system =
+                    capnp_rpc::RpcSystem::new(Box::new(network), Some(client.clone().client));
+
+                tokio::task::spawn_local(rpc_system);
+            });
+
+            while let Ok(message) = output_receiver.recv() {
+                messages_callback.as_mut().call(&message);
+            }
+        });
+
+        Ok(Impl {
+            sender: input_sender,
+            write_shutdown: Arc::new(std::sync::atomic::AtomicBool::new(false)),
+        })
+    }
+}
+
+struct Server {
+    docker: Arc<Docker>,
+    container_name: String,
+}
+
+impl Server {
+    fn connect(address: &str, container_name: &str) -> Result<Self, ContainerError> {
+        let docker = Docker::connect_with_socket(address, 120, API_DEFAULT_VERSION)
+            .map_err(|e| ContainerError::ConnectionError(e.to_string()))?;
+        Ok(Server {
+            docker: docker.into(),
+            container_name: container_name.to_owned(),
+        })
+    }
+}
+
+impl container::Server for Server {
+    fn destroy(
+        &mut self,
+        _: container::DestroyParams,
+        _: container::DestroyResults,
+    ) -> Promise<(), capnp::Error> {
+        let container_name = self.container_name.clone();
+        let docker = self.docker.clone();
+        Promise::from_future(async move {
+            let options = StopContainerOptionsBuilder::default().signal("SIGINT");
+            docker
+                .stop_container(&container_name, Some(options.build()))
+                .await
+                .map_err(|e| ContainerError::DestroyError(e.to_string()))?;
+            Ok(())
+        })
+    }
+
+    fn signal(
+        &mut self,
+        params: container::SignalParams,
+        _: container::SignalResults,
+    ) -> Promise<(), capnp::Error> {
+        let container_name = self.container_name.clone();
+        let docker = self.docker.clone();
+        Promise::from_future(async move {
+            let signo = params.get()?.get_signo();
+            let signal_name = match signo {
+                1 => "SIGHUP",
+                2 => "SIGINT",
+                3 => "SIGQUIT",
+                9 => "SIGKILL",
+                15 => "SIGTERM",
+                18 => "SIGCONT",
+                19 => "SIGSTOP",
+                20 => "SIGTSTP",
+                _ => "SIGTERM", // Default fallback
+            };
+
+            let options = StopContainerOptionsBuilder::default().signal(signal_name);
+            docker
+                .stop_container(&container_name, Some(options.build()))
```

Comment:
I assume `stop_container` is just kind of a misnomer from the library we're using?

Reply:
What should destroy do? should it also remove the container?

Reply:
I've updated the implementation to also call removeContainer(). Let me know if it's incorrect.

Reply:
Commented inline on destroy. - I think this comment was on `signal` not `destroy` FWIW! Hence me asking if stop_container was a misnomer.